samples/file-encryptor/enclave/openssl/keys.cpp

```diff
@@ -0,0 +1,428 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include <openssl/rand.h>
+#include <string.h>
+#include "../common/common.h"
+#include "encryptor.h"
+
+#define ENCRYPT_OPERATION true
+#define DECRYPT_OPERATION false
+#define SALT_SIZE_IN_BYTES 16 // Length of salt size
+
+void ecall_dispatcher::dump_data(
+    const char* name,
+    unsigned char* data,
+    size_t data_size)
+{
+    TRACE_ENCLAVE("Data name: %s", name);
+    for (size_t i = 0; i < data_size; i++)
+    {
+        TRACE_ENCLAVE("[%ld]-0x%02X", i, data[i]);
+    }
+    TRACE_ENCLAVE("\n");
+}
+
+// Compute the sha256 hash of given data.
+int ecall_dispatcher::Sha256(
+    const uint8_t* data,
+    size_t data_size,
+    uint8_t sha256[32])
+{
+    int ret = -1;
+    unsigned int sha_length_in_bytes = 32;
+    EVP_MD_CTX* mdctx;
+
+    if ((mdctx = EVP_MD_CTX_new()) == NULL)
+    {
+        TRACE_ENCLAVE("Message digest context instatntiation failed");
+        goto exit;
+    }
+
+    if (1 != EVP_DigestInit_ex(mdctx, EVP_sha256(), NULL))
+    {
+        TRACE_ENCLAVE(
+            "Message digest context initialization to SHA-256 failed");
+        goto exit;
+    }
+
+    if (1 != EVP_DigestUpdate(mdctx, data, data_size))
+    {
+        TRACE_ENCLAVE("Message digest context update to input data failed");
+        goto exit;
+    }
+
+    if (1 != EVP_DigestFinal_ex(mdctx, sha256, &sha_length_in_bytes))
+    {
+        TRACE_ENCLAVE("Message Digest Final failed");
+        goto exit;
+    }
+
+    ret = 0;
+    EVP_MD_CTX_free(mdctx);
+exit:
+    return ret;
+}
+
+// This routine uses the openssl library to derive an AES key from the input
+// password and produce a password based key. Note : A set of hardcoded salt
+// values are used here for the purpose simplifying this sample, which caused
+// this routine to return the same key when taking the same password. This saves
+// the sample from having to write salt values to the encryption header. In a
+// real world application, randomly generated salt values are recommended.
```

Comment:
non-blocking comment: experience has shown that people copy sample code into actual products.   As such, samples should do what is recommended, not do what is recommended against.  So this sample should be updated (possibly in a subsequent PR) to use randomly generated salt values so that it does not encourage security bugs to creep into products.

Reply:
Created https://github.com/openenclave/openenclave/issues/3692 to track this